### PR TITLE
Patch boringssl for ppc64le

### LINKED ISSUE #20677: C++ compilation of rule '@boringssl//:crypto' failed (Exit 1)
[jingfeng@Jingfeng_redhat tensorflow]$ bazel build --config=opt --config=cuda //tensorflow/tools/pip_package:build_pip_package

ERROR: /home/jingfeng/.cache/bazel/_bazel_jingfeng/29120beece5f660cdcebf72073ed1bba/external/boringssl/BUILD:115:1: C++ compilation of rule '@boringssl//:crypto' failed (Exit 1)
external/boringssl/src/crypto/cipher_extra/e_chacha20poly1305.c: In function 'poly1305_update_length':
external/boringssl/src/crypto/cipher_extra/e_chacha20poly1305.c:145:3: error: 'for' loop initial declarations are only allowed in C99 mode
for (unsigned i = 0; i < sizeof(length_bytes); i++) {
^
external/boringssl/src/crypto/cipher_extra/e_chacha20poly1305.c:145:3: note: use option -std=c99 or -std=gnu99 to compile your code
external/boringssl/src/crypto/cipher_extra/e_chacha20poly1305.c: In function 'aead_chacha20_poly1305_seal_scatter':
external/boringssl/src/crypto/cipher_extra/e_chacha20poly1305.c:229:5: error: 'for' loop initial declarations are only allowed in C99 mode
for (size_t done = 0; done < extra_in_len; block_counter++) {
^
external/boringssl/src/crypto/cipher_extra/e_chacha20poly1305.c:233:7: error: 'for' loop initial declarations are only allowed in C99 mode
for (size_t i = offset; i < sizeof(block) && done < extra_in_len;
^
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 8.244s, Critical Path: 6.30s
INFO: 5 processes: 5 local.
FAILED: Build did NOT complete successfully

[jingfeng@Jingfeng_redhat ~]$ uname -a
Linux Jingfeng_redhat 3.10.0-862.3.2.el7.ppc64le #1 SMP Tue May 15 22:31:16 UTC 2018 ppc64le ppc64le ppc64le GNU/Linux

[jingfeng@Jingfeng_redhat ~]$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/ppc64le-redhat-linux/4.8.5/lto-wrapper
Target: ppc64le-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-ppc64le-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-ppc64le-redhat-linux/cloog-install --enable-gnu-indirect-function --enable-secureplt --with-long-double-128 --enable-targets=powerpcle-linux --disable-multilib --with-cpu-64=power8 --with-tune-64=power8 --build=ppc64le-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-28) (GCC)

jingfeng@Jingfeng_redhat ~]$ bazel version
Extracting Bazel installation...
Build label: 0.11.0- (@non-git)
Build target: bazel-out/ppc-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Jul 13 14:25:29 +50491 (1531191651929)
Build timestamp: 1531191651929
Build timestamp as int: 1531191651929

python3.4 
cuda9.0 
cudnn latest

[root@Jingfeng_redhat ~]# nvidia-smi
Tue Jul 10 17:50:17 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 396.26                 Driver Version: 396.26                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 00000000:03:00.0 Off |                  Off |
| N/A   30C    P8    26W / 149W |      0MiB / 12206MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           On   | 00000000:04:00.0 Off |                  Off |
| N/A   25C    P8    30W / 149W |      0MiB / 12206MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           On   | 00000002:03:00.0 Off |                  Off |
| N/A   41C    P0    N/A /  N/A |      0MiB / 12206MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           On   | 00000002:04:00.0 Off |                  Off |
| N/A   26C    P8    30W / 149W |      0MiB / 12206MiB |      0%      Default |

May I know the reason for this problem? & how to solve it? Many thanks for this

### PR DESCRIPTION
**Pull request by wdirons**
This commit adds a patch file for the bazel BUILD file contained in the
boringssl.tar.gz file. It set the necessary compile options to build boringssl
on pp64le. (Basically adds linux_ppc64le everywhere linux_x86_64 is)

Fixes #20677

### CODE REVIEW

**Comment by wdirons** [comment]
Related question , if I wanted to upstream this fix so the patch wouldn't be required for Tensorflow where would I do it? The tar.gz file that is downloaded for boringssl includes more files then what I get when cloning https://boringssl.googlesource.com/boringssl/

**Comment by wdirons** [comment]
I found that the boringssl repo has a master-with-bazel branch that I need to work from.

File: third_party/boringssl/boringssl_build_file_ppc64le.patch [hunk 1]
```diff
@@ -0,0 +1,53 @@
+--- a/BUILD     2017-10-17 21:30:48.000000000 +0000
++++ b/BUILD     2018-01-09 12:40:36.264153992 +0000
+@@ -23,6 +23,7 @@
+     "crypto_sources",
+     "crypto_sources_linux_x86_64",
+     "crypto_sources_mac_x86_64",
++    "crypto_sources_linux_ppc64le",
```
Comment (gunan) [inline third_party/boringssl/boringssl_build_file_ppc64le.patch]: Previously when we  asked boringssl about such patches, they told us some of these may not work at all.
Can we work with them to make sure this works, and then upstream it into boringssl?

**Comment by jayfurmanek** [comment]
@gunan Can we get a quick re-review? Looks like @wdirons had to make one last change to resolve https://github.com/tensorflow/tensorflow/issues/20677

Also is it possible to pick this over to the 1.10 branch?

Thx!

**Review by gunan — APPROVED** [review]

**Pull request merged**